ca-certificates 2021-07-05 (new formula)

Bo98 opened the pull request:
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Also, stop using the keychain on Yosemite and El Capitan, since these versions don't have the Let's Encrypt root.

I have not tested this on a 10.10 VM yet, but will before merging to make sure the PEM route actually works.

carlocab commented on Formula/openssl@1.1.rb:
```diff
@@ -25,8 +25,9 @@ class OpensslAT11 < Formula
 
   on_linux do
     resource "cacert" do
```
Let's just make this a separate formula. It's in use in multiple formulae anyway.

Bo98 replied:
Not sure of the best strategy for that.

Ideally we'd handle the keychain stuff there and then the file can be used by multiple formula. But the problem is we currently use `openssl` in that processing. A bit of a chicken and egg problem.

Bo98 replied:
Or maybe `/usr/bin/openssl` is enough actually. I forget that still exists.

carlocab replied:
I don't follow -- the `cacert` resource is used only in `post_install_from_pem`, where no `openssl` binary is needed. Where do we need `openssl` here?

Or, I guess alternatively we can just use `/etc/ssl/cert.pem`. Is there any reason why we don't use that one?

Bo98 replied:
> I don't follow -- the `cacert` resource is used only in `post_install_from_pem`, where no `openssl` binary is needed. Where do we need `openssl` here?

The keychain part. I've just pushed what I was thinking of.

> Or, I guess alternatively we can just use `/etc/ssl/cert.pem`

I can't recall the differences between that and the keychain, but we're going to want to avoid that on older macOS anyway since it's severely outdated.

carlocab commented on Formula/openssl@1.1.rb:
```diff
@@ -23,14 +23,9 @@ class OpensslAT11 < Formula
-      mirror "https://gist.githubusercontent.com/dawidd6/16d94180a019f31fd31bc679365387bc/raw/ef02c78b9d6427585d756528964d18a2b9e318f7/cacert-2020-01-01.pem"
-      sha256 "adf770dfd574a0d6026bfaa270cb6879b063957177a991d453ff1d302c02081f"
-    end
+  depends_on "cacert"
```
I was thinking of making this a Linux- and old-macOS-only dependency, so that we can keep the keychain `post_install` in this formula. This way we can still do `depends_on "cacert"` in `gnutls`.

Bo98 replied:
gnutls does the exact same Keychain stuff. I'm not sure I follow how there wouldn't be a conditional in `gnutls` as well.

carlocab replied:
It's not exactly the same:

https://github.com/Homebrew/homebrew-core/blob/c30079a98e07b0138f7cb2cce418cb8d394f1ae4/Formula/gnutls.rb#L104-L112

Bo98 replied:
Honestly, I see no reason why that can't apply to OpenSSL as well. It's a self-signed Kerberos certificate - not a SSL root certificate.

carlocab commented on Formula/cacert.rb:
```diff
@@ -0,0 +1,87 @@
+    end
+  end
+
+  def post_install_from_keychain
```
It's kinda weird to do this here. On new macOS, you install `cacert` for the `post_install`, when you don't even actually use anything it downloads.

carlocab commented on Formula/openssl@1.1.rb:
```diff
@@ -123,64 +118,9 @@ def openssldir
-    # Download and install cacert.pem from curl.haxx.se
-    cacert = resource("cacert")
-    cacert.fetch
+    Formula["cacert"].post_install
```
Maybe add an
```ruby
ENV.prepend_path "PATH", opt_bin
```
so that the `Formula["cacert"].post_install` uses this formula's `openssl`? Not sure it makes a difference, but that's what this formula uses right now...

Bo98 replied:
I could do, but that introduces inconsistency when `cacert` is updated where it'll go back to using `/usr/bin/openssl`.

I doubt there's much differences since all it is used for is for reading the issuer and expiry date.

Bo98 commented:
The general idea I have here is I would also like to get rid of the copy paste chaos we have where I have to make the same change in 3 different places. Plus decoupling CA cert updates from an OpenSSL & GnuTLS revision bump is nice.

carlocab commented on Formula/cacert.rb:
```diff
@@ -0,0 +1,87 @@
+    end
+
+    pkgetc.mkpath
+    (pkgetc/"cert.pem").atomic_write(trusted_certs.join("\n") << "\n")
```
Is there any reason not to add the curl.se `cert.pem` here? May as well make use of the extra stuff you now download with this formula.

Bo98 replied:
Hmm maybe. I'll compare them and see what the current differences are.

Bo98 replied:
I'll get this to you tomorrow. Just realised it's 4.30am 😅

Bo98 replied:
https://gist.github.com/Bo98/3717e3384442e454b38be0259d094e4d
https://gist.github.com/Bo98/e958f5b3c2a23ed338de94e0a3b6e248

Bo98 replied:
(Note that some differences will be because Apple are slower to remove expired certs)

Bo98 replied:
Added

Bo98 commented:
Oh and to resolve #54235 finally.

carlocab commented:
> Oh and to resolve #54235 finally.

@FiloSottile will be pleased to learn about this. (Apologies for the ping, but this stuff seems to be in your wheelhouse, so it'd be nice to hear your thoughts here too.)

carlocab commented on Formula/cacert.rb:
```diff
@@ -0,0 +1,88 @@
+class Cacert < Formula
```
```suggestion
class CaCertificates < Formula
```
Maybe?

Bo98 replied:
I'm fine with that. I have no strong feelings on the name.

Bo98 commented:
> @FiloSottile will be pleased to learn about this.

Just to summarise, the idea is:

`HOMEBREW_PREFIX/etc/cacert/cert.pem` is the only copy on the system (or whatever we call the formula).

`HOMEBREW_PREFIX/etc/openssl@1.1/cert.pem` is a symlink to the above. Same with those in `gnutls`, `openssl@3` etc.

carlocab commented on Formula/openssl@1.1.rb:
```diff
@@ -123,64 +118,9 @@ def openssldir
     rm_f openssldir/"cert.pem"
-    filename = Pathname.new(cacert.url).basename
-    openssldir.install cacert.files(filename => "cert.pem")
+    ln_s "../cacert/cert.pem", openssldir/"cert.pem"
```
```suggestion
    openssldir.install_symlink Formula["cacert"].pkgetc/"cert.pem"
```
`install_symlink` should resolve the correct relative path here since these paths don't live in different kegs.

Bo98 commented:
@Homebrew/core I know this topic has come up on a number of occasions on Slack so here's a ping if you're interested (and sorry if you're not).

SMillerDev commented on Formula/cacert.rb:
```diff
@@ -0,0 +1,88 @@
+  desc "Mozilla CA certificate store"
+  homepage "https://curl.se/docs/caextract.html"
+  url "https://curl.se/ca/cacert-2021-07-05.pem"
+  # Ideally we'd have a HTTP mirror, but a GitHub-hosted one is second best.
+  mirror "https://gist.githubusercontent.com/jonchang/b13c60cd50bdd91fd12a034ea53e39d5/raw/e1c430fd9ab59ed34212083537be0da974c26555/cacert-2021-07-05.pem"
```
Can we just put it in `formula-patches`?

Bo98 commented:
Latest push:

* Removed tempfile usage - fixes https://github.com/Homebrew/brew/issues/11247
* Fix a large process descriptor leak (discovered in 10.10 VM - which has low process limits)
* Fix a security issue where `verify-cert` step introduced in #71191 wasn't actually doing anything
  - This makes the Kerberos/GnuTLS workaround no longer necessary, though I've kept it there for now in case someone has manually trusted it.

All of these issues were pre-existing in `openssl@1.1` postinstall.

carlocab commented on Formula/openssl@1.1.rb:
```diff
@@ -23,14 +23,9 @@ class OpensslAT11 < Formula
-      mirror "https://gist.githubusercontent.com/dawidd6/16d94180a019f31fd31bc679365387bc/raw/ef02c78b9d6427585d756528964d18a2b9e318f7/cacert-2020-01-01.pem"
-      sha256 "adf770dfd574a0d6026bfaa270cb6879b063957177a991d453ff1d302c02081f"
-    end
+  depends_on "ca-certificates"
```
Let's bump the `revision` here. Will we need to stop setting `HOMEBREW_FORCE_BREWED_CURL` on Mojave for things to work here?

Bo98 replied:
> Will we need to stop setting `HOMEBREW_FORCE_BREWED_CURL` on Mojave for things to work here?

Yes.

Was hoping to bump the revision at the same time as making it keg-only, but that's going to be a bit of work since OpenSSL dependent testing is busted right now.

FiloSottile commented:
Thank you for the ping, consolidating the root store logic is a great idea 👍

I’m not sure I understand the choice of merging the Mozilla and Apple root stores. Is it just to bring in the Let’s Encrypt root on older macOS versions? Could it be scoped to only systems that are not getting updates from Apple? Root stores are not a case of more is better :)

I did a quick skim of the logic, but it seems to ignore the root purposes: in a store you can have roots trusted for email, code signing, Kerberos, TLS… you usually only want to extract the latter. Trusting the others can be a significant security liability. security(1) has a flag to pick the “policy”, and you want “SSL”. This should also avoid the Kerberos special case, which is a red flag. You can look at how we used to do it in Go at https://go-review.googlesource.com/c/go/+/227037/17/src/crypto/x509/root_darwin.go. The Mozilla store also has purposes, but what you get depends on where you download it. Maybe curl did the filtering for you.

An important note is that the formula revision should be bumped when Apple updates the root store, so it stays in sync.

carlocab commented on Formula/openssl@1.1.rb:
```diff
@@ -23,14 +23,9 @@ class OpensslAT11 < Formula
-    end
+  depends_on "ca-certificates"
 
+  on_linux do
```
This can probably be dealt with in a separate PR, but I wonder about these `resource`s. Does OpenSSL really need these at runtime? These look like build dependencies to me.

Bo98 replied:
No, but I think there's some dependents that are. This should be moved elsewhere but that's going to be out of scope for this PR.

Bo98 commented:
> I’m not sure I understand the choice of merging the Mozilla and Apple root stores. Is it just to bring in the Let’s Encrypt root on older macOS versions?

Yes.

> Could it be scoped to only systems that are not getting updates from Apple?

Apple seem to rarely update older trust stores. The Mojave cert store, for example, has not been modified since the 2019-002 security update (December 2019). Catalina's cert store has not been modified since June 2020 (presumably 10.15.6). Monterey (beta) meanwhile was updated a month ago.

So "not getting update from Apple" might be anything that's not the latest version.

> I did a quick skim of the logic, but it seems to ignore the root purposes: in a store you can have roots trusted for email, code signing, Kerberos, TLS… you usually only want to extract the latter.

Yep I absolutely agree. This should have been added when the keychain was expanded beyond Apple roots. I've added a restriction to the "SSL Client CA" and "SSL Server CA" purpose.

`-p ssl` would probably also work with `verify-cert` but apparently that starts making network calls.

> The Mozilla store also has purposes, but what you get depends on where you download it. Maybe curl did the filtering for you.

Curl store filters to only include `SERVER_AUTH`.

FiloSottile commented:
> Yep I absolutely agree. This should have been added when the keychain was expanded beyond Apple roots. I've added a restriction to the "SSL Client CA" and "SSL Server CA" purpose.

You probably want only "SSL Server CA", which are the ones actually bound by the Baseline Requirements. (It also matches curl's `SERVER_AUTH`.)

Bo98 commented:
> > Yep I absolutely agree. This should have been added when the keychain was expanded beyond Apple roots. I've added a restriction to the "SSL Client CA" and "SSL Server CA" purpose.
> 
> You probably want only "SSL Server CA", which are the ones actually bound by the Baseline Requirements. (It also matches curl's `SERVER_AUTH`.)

Yeah kinda just realised after pushing. Fixed now!

carlocab commented on Formula/openssl@1.1.rb:
```diff
@@ -123,64 +119,9 @@ def openssldir
-    # Download and install cacert.pem from curl.haxx.se
-    cacert = resource("cacert")
-    cacert.fetch
+    Formula["ca-certificates"].post_install
```
Won't this make everyone have the `post_install` done twice when they `brew upgrade` or `brew install openssl@1.1` once this is merged? I can see why we need it, but I wonder if there's a way to avoid doing this twice in a single `brew` invocation.

Bo98 replied:
I can probably drop this in favour of regular `ca-certificates` updates & revision bumps, which is what we should be doing.

carlocab commented:
Just noticed we used to ship a CA bundle: https://github.com/Homebrew/legacy-homebrew/pull/28658

This doesn't necessarily have any bearing on what we're doing here today, but I thought I'd link this here for anyone who needs to look up things we've done before and the reasons for them.

Bo98 commented:
So many chickens, so many eggs.

10.14 audit of course fails because it needs https://github.com/Homebrew/brew/pull/12167, but https://github.com/Homebrew/brew/pull/12167 in turn needs this PR.

Anyhow, this should be good to go.

carlocab commented:
🐔 🥚 🐔

carlocab commented:
Not an `all` bottle. Very strange.

carlocab commented:
Timestamps are off:
```
❯ diffoscope ca-certificates--2021-09-30.arm64_big_sur.bottle.tar.gz ca-certificates--2021-09-30.mojave.bottle.tar.gz
--- ca-certificates--2021-09-30.arm64_big_sur.bottle.tar.gz
+++ ca-certificates--2021-09-30.mojave.bottle.tar.gz
├── filetype from file(1)
│ @@ -1 +1 @@
│ -gzip compressed data, was "ca-certificates-bottle.tar", last modified: Thu Sep 30 21:42:41 2021, from Unix
│ +gzip compressed data, was "ca-certificates-bottle.tar", last modified: Mon Oct  4 02:43:56 2021, from Unix
│   --- ca-certificates--2021-09-30.arm64_big_sur.bottle.tar
├── +++ ca-certificates--2021-09-30.mojave.bottle.tar
│ ├── file list
│ │ @@ -1,6 +1,6 @@
│ │ -drwxr-xr-x   0        0        0        0 2021-09-30 21:42:41.000000 ca-certificates/2021-09-30/
│ │ -drwxr-xr-x   0        0        0        0 2021-09-30 21:42:41.000000 ca-certificates/2021-09-30/.brew/
│ │ --rw-r--r--   0        0        0     4226 2021-09-30 21:42:41.000000 ca-certificates/2021-09-30/.brew/ca-certificates.rb
│ │ -drwxr-xr-x   0        0        0        0 2021-09-30 21:42:41.000000 ca-certificates/2021-09-30/share/
│ │ -drwxr-xr-x   0        0        0        0 2021-09-30 21:42:41.000000 ca-certificates/2021-09-30/share/ca-certificates/
│ │ --rw-r--r--   0        0        0   203007 2021-09-30 21:42:41.000000 ca-certificates/2021-09-30/share/ca-certificates/cacert.pem
│ │ +drwxr-xr-x   0        0        0        0 2021-10-04 02:43:56.000000 ca-certificates/2021-09-30/
│ │ +drwxr-xr-x   0        0        0        0 2021-10-04 02:43:56.000000 ca-certificates/2021-09-30/.brew/
│ │ +-rw-r--r--   0        0        0     4226 2021-10-04 02:43:56.000000 ca-certificates/2021-09-30/.brew/ca-certificates.rb
│ │ +drwxr-xr-x   0        0        0        0 2021-10-04 02:43:56.000000 ca-certificates/2021-09-30/share/
│ │ +drwxr-xr-x   0        0        0        0 2021-10-04 02:43:56.000000 ca-certificates/2021-09-30/share/ca-certificates/
│ │ +-rw-r--r--   0        0        0   203007 2021-10-04 02:43:56.000000 ca-certificates/2021-09-30/share/ca-certificates/cacert.pem
```

Bo98 commented:
That is really weird and makes no sense. I could try a rebottle? It definitely was working in one of the attempts here, and I've only been fiddling with postinstall since.

carlocab commented:
I'll poke at it; go to bed.

MikeMcQuaid commented on Formula/ca-certificates.rb:
```diff
@@ -0,0 +1,124 @@
+
+  def post_install
+    if OS.mac?
+      macos_post_install
+    else
+      linux_post_install
```
I'd be tempted to just inline these.

MikeMcQuaid reviewed (commented):
Great work here!
